cardinal/storage/redis/nonce.go

```diff
@@ -3,35 +3,140 @@ package redis
+	mutex      *sync.Mutex
+	maxNonce   map[string]uint64
+	countNonce map[string]int
 }
 
 func NewNonceStorage(client *redis.Client) NonceStorage {
 	return NonceStorage{
-		Client: client,
+		Client:     client,
+		mutex:      &sync.Mutex{},
+		maxNonce:   map[string]uint64{},
+		countNonce: map[string]int{},
 	}
 }
 
-var ErrNonceHasAlreadyBeenUsed = errors.New("nonce has already been used")
-
 // UseNonce atomically marks the given nonce as used. The nonce is valid if nil is returned. A non-nil error means
 // there was an error verifying the nonce, or the nonce was already used.
 func (r *NonceStorage) UseNonce(signerAddress string, nonce uint64) error {
+	if nonce > maxValidNonce {
+		return eris.New("nonce is too large")
+	}
 	ctx := context.Background()
 	key := r.nonceSetKey(signerAddress)
-	added, err := r.Client.SAdd(ctx, key, nonce).Result()
+
+	// All redis and in-memory map operations happen inside a lock. This could be improved by creating a separate lock
+	// for each signer address.
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+
+	maxNonce, err := r.getMaxNonceForKey(ctx, key)
+	if err != nil {
+		return eris.Wrap(err, "failed to get max nonce for signer address")
+	}
+
+	// Nonces beyond the sliding window are invalid and can be rejected outright.
+	if nonce < maxNonce && maxNonce-nonce >= NonceSlidingWindowSize {
+		return eris.New("nonce is too old")
+	}
+
+	zItem := redis.Z{
+		Score:  float64(nonce),
+		Member: nonce,
+	}
+
+	added, err := r.Client.ZAdd(ctx, key, zItem).Result()
 	if err != nil {
-		return err
+		return eris.Wrap(err, "failed to add nonce")
 	}
-	// The nonce was already used
+	// A result of 0 from ZAdd means no new items were actually added to the Zset. This means the nonce was already
+	// used.
 	if added == 0 {
 		return eris.Wrapf(ErrNonceHasAlreadyBeenUsed, "signer %q has already used nonce %d", signerAddress, nonce)
 	}
+
+	r.maxNonce[key] = max(r.maxNonce[key], nonce)
+	r.countNonce[key]++
+
+	if r.countNonce[key] > numOfNoncesToTriggerCleanup {
+		r.cleanupOldNonces(ctx, key, r.maxNonce[key])
+	}
+
 	return nil
 }
+
+// cleanupOldNonces removes the record of all nonces that are older than NonceSlidingWindowSize. Nonces in that range
+// can be rejected without checking storage. ZRemRangeByScore has a performance of O(log(N)+M) where N is the number
+// of items in the set and M is the number of items to remove.
+func (r *NonceStorage) cleanupOldNonces(ctx context.Context, key string, currMax uint64) {
+	minScore := "-inf"
+	maxScore := fmt.Sprintf("%d", currMax-NonceSlidingWindowSize)
+	removed, err := r.Client.ZRemRangeByScore(ctx, key, minScore, maxScore).Result()
+	if err != nil {
+		log.Err(err).Msg("failed to remove old nonces")
+		return
+	}
+	r.countNonce[key] -= int(removed)
+}
+
+// getMaxNonceForKey returns the highest used nonce for the given key.
+func (r *NonceStorage) getMaxNonceForKey(ctx context.Context, key string) (uint64, error) {
```

Comment:
Should we be more specific and clarify that the `key` is in fact just a `signerAddress`? Seems like that is the only possible thing we would use here so being specific and clear makes sense as opposed to leaving it generic?

Reply:
Updated to signerAddressKey. It's not just the signer address, but also common prefix to help identify all these keys: `USED_NONCES_<some-signer-address>`